File: Core-backend-layer/Programming-fundamentals/Python/OOP/src/nested_classes/product.py
class Product:
    global_review_count = 0  # Class-level variable shared among all instances

    def __init__(self, product_name, price, available_quantity):
        self.product_name = product_name
        self.price = price
        self.available_quantity = available_quantity

    def display_info(self):
        print(f"Product: {self.product_name}, Price: ${self.price}, Available quantity: {self.available_quantity}")

    # Nested class - in Python, there is no explicit distinction between a static nested class and a member class
    class Discount:
        def __init__(self, product_instance, discount_percentage):
            # In python, the outer class instance needs to be explicitly passed to the inner class constructor
            self.product_instance = product_instance
            self.discount_percentage = discount_percentage

        def apply_discount(self):
            # Accessing the price variable of the outer class using the provided instance
            discounted_price = self.product_instance.price - (self.product_instance.price * (self.discount_percentage / 100))
            print(f"Applying a {self.discount_percentage}% discount. Discounted price: ${discounted_price}")

    # Unlike Java, Python doesn't have a separate keyword for static nested classes
    class Review:
        review_count = 0

        def __init__(self, product_instance, comment, rating):
            # In python, the outer class instance needs to be explicitly passed to the inner class constructor
            self.product_instance = product_instance
            self.comment = comment
            self.rating = rating
            Product.Review.review_count += 1
            Product.global_review_count += 1

        def display_review(self):
            print(f"Review: {self.comment}, Rating: {self.rating}, Total number of reviews: {Product.global_review_count}")

        # Decorator for the method to be associated with the class rather than the instance
        @staticmethod
        def get_review_count():
            return Product.Review.review_count

        @staticmethod
        def get_global_review_count():
            return Product.global_review_count

    def get_available_quantity(self):
        return self.available_quantity

    def purchase(self, quantity):
        if quantity <= self.available_quantity:
            print(f"Purchased {quantity} units of {self.product_name}")
            self.available_quantity -= quantity
        else:
            print(f"Insufficient quantity in stock for {self.product_name}")

File: Core-backend-layer/Programming-fundamentals/Java/OOP/src/InnerClasses/Product.java
public class Product {
    private String productName;
    private double price;
    private int availableQuantity;
    // Static variables are class-level properties, shared among all instances, with only one copy for the entire class.
    private static int globalReviewCount = 0;

    public Product(String productName, double price, int availableQuantity) {
        this.productName = productName;
        this.price = price;
        this.availableQuantity = availableQuantity;
    }

    void displayInfo() {
        System.out.println("Product: " + productName + ", Price: $" + price + ", Available quantity: " + availableQuantity);
    }

    // Member class for applying discounts
    class Discount {
        private double discountPercentage;

        public Discount(double discountPercentage) {
            this.discountPercentage = discountPercentage;
        }

        // Applying a discount using both outer and inner class variables, member class has access to outer class's instance variables
        void applyDiscount() {
            // Accessing the price variable of the outer class directly.
            double discountedPrice = price - (price * (discountPercentage / 100));
            System.out.println("Applying a " + discountPercentage + "% discount. Discounted price: $" + discountedPrice);
        }
    }

    // Static nested class for handling product reviews
    static class Review {
        private String comment;
        private int rating;
        // A static variable in an inner class is associated with the inner class, and it is not directly related to instances of the outer class.
        private static int reviewCount;

        public Review(String comment, int rating) {
            this.comment = comment;
            this.rating = rating;
            reviewCount++;
            globalReviewCount++;
        }

        // Static nested classes do not have access to the instance variables and methods of the outer class unless they are static.
        public void displayReview() {
            System.out.println("Review: " + comment + ", Rating: " + rating + ", Total number of reviews: " + globalReviewCount);
        }

        public static int getReviewCount() {
            return reviewCount;
        }

        public static int getGlobalReviewCount() {
            return globalReviewCount;
        }
    }

    public int getAvailableQuantity() {
        return availableQuantity;
    }

    public void purchase(int quantity) {
        if (quantity <= availableQuantity) {
            System.out.println("Purchased " + quantity + " units of " + productName);
            availableQuantity -= quantity;
        } else {
            System.out.println("Insufficient quantity in stock for " + productName);
        }
    }
}